src/flat_list.jsx

```diff
@@ -1,71 +1,154 @@
 
 const itemChecker = (prevProps, nextProps) => {
   return isEqual(prevProps, nextProps);
 };
 
-const ItemRenderer = React.memo(({ item, renderItem }) => renderItem(item), itemChecker);
+class ItemRenderer extends React.Component {
+  shouldComponentUpdate() {
+    return !itemChecker();
+  }
+  render() {
+    const { item, renderItem } = this.props;
+    return renderItem(item);
+  }
+}
 
 class FlatList extends React.Component {
   constructor(props) {
     super(props);
     this.state = {
       items: []
     };
+    this.getLastElementHeight = this.getLastElementHeight.bind(this);
+    this.getKey = this.getKey.bind(this);
+    this.loadMoreItems = this.loadMoreItems.bind(this);
+    this.scrollToItem = this.scrollToItem.bind(this);
   }
 
   componentDidMount() {
-    const { data, batchCount, isVirtual } = this.props;
-    if(isVirtual) {
-      this.setState({
-        items: data.slice(0, batchCount)
+    const { data, batchCount, autoLoad, loadOnScroll, scrollParent } = this.props;
+    const { items } = this.state;
+
+    if (autoLoad) {
+      if (data.length > 0) {
+        this.setState({
+          items: data.slice(0, batchCount)
+        });
+      }
+    } else if (loadOnScroll) {
+      if (data.length > 0 && items.length == 0) {
+        this.setState({
+          items: data.slice(0, batchCount)
+        });
+      }
+      const parent = scrollParent ? document.querySelector(scrollParent) : window;
+      const parentHeight = scrollParent ? parent.clientHeight : window.innerHeight;
+      parent.addEventListener("scroll", () => {
+        if (this.getLastElementHeight() - 300 <= parentHeight) {
+          this.loadMoreItems();
+        }
       });
     }
   }
 
   componentDidUpdate({ data: oldData }) {
-    const { data, batchCount, isVirtual } = this.props;
+    const { data, batchCount, autoLoad, loadOnScroll, positionToScroll } = this.props;
     const { items } = this.state;
 
-    if (isVirtual) {
+    //---------- To Load Items Automatically -----------//
+    if (autoLoad) {
       if (oldData.length == data.length && itemChecker({ data: oldData }, { data })) {
-        clearTimeout(this.timeoutId);
-        const hasMore =
-          items.length + batchCount <= data.length ||
-          (data.length - items.length <= batchCount && data.length !== items.length);
-        if (hasMore) {
-          this.timeoutId = setTimeout(() => {
-            this.setState(prevState => ({
-              items: [...items, ...data.slice(prevState.items.length, prevState.items.length + batchCount)]
-            }));
-          }, timeoutMilliSec);
-        }
+        this.loadMoreItems();
+      } else {
+        this.setState({ items: [] });
+      }
+    }
+
+    //---------- To Load first batch of Items if items length is 0 before scrolling ---------//
+    else if (loadOnScroll) {
+      if (oldData.length == data.length && itemChecker({ data: oldData }, { data })) {
+        if (items.length == 0)
```

Comment:
This part has been used multiple times. Why Dont we extract to a common instance function ?